### PR TITLE
Support conversion between Python NumPy arrays, Matlab matrices, and Simbody numeric types

### LINKED ISSUE #1882: Convert OpenSim numeric types to MATLAB matrices
Converting OpenSim Vectors into MATLAB matrices is challenging but is a common task.

MATLAB supports converting java arrays to matrices:

```
cell2mat(cell(javaArray))
```

We already have a converter from `Vec3` to a java array here: https://github.com/opensim-org/opensim-core/blob/master/Bindings/Java/swig/java_simbody.i#L20.

This means that one can do the following right now in MATLAB:

```
cell2mat(cell(vec3.getAsJavaArray()))
```

This is still very verbose, but perhaps we could create a utility function `osimMakeMatrix()` or something like that.

@tkuchida and I just looked into this together.

Links:

- https://www.mathworks.com/help/matlab/matlab_external/access-elements-of-java-array.html
- https://www.mathworks.com/help/matlab/ref/javaarray.html
- https://www.mathworks.com/help/matlab/matlab_external/how-matlab-represents-java-array.html
- https://www.mathworks.com/matlabcentral/newsreader/view_thread/169936

### PR DESCRIPTION
**Pull request by chrisdembia**
Fixes #1882 and addresses part of #2559.

### Brief summary of changes

- Allow converting between Python NumPy arrays and common Simbody numeric types.
- Allow converting between Matlab matrices and common Simbody numeric types.
- NumPy is now required when building the Python bindings; I updated the continuous integration scripts accordingly.
- This PR was originally motivated by Moco's dependence on NumPy.
- No support yet for conversions with (Row)VectorVec3 in Python.

### Testing I've completed

Tested all new functions in `testSimbody.m` and `test_simbody.py`.

### Looking for feedback on...

Names of new functions; `createFromMat()` and `getAsMat()`. These names are consistent with our use of `Mat` in Moco for similar conversions.

### CHANGELOG.md (choose one)

- updated...

The Doxygen for this PR can be viewed at http://myosin.sourceforge.net/?C=N;O=D; click the folder whose name is this PR's number.

---
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/2706)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -506,6 +506,7 @@ On Windows using Visual Studio
         * Note: Java development kit >= 9 requires CMake >= 3.10.
     * **Python scripting** (optional): Python 2 >= 2.7 or Python 3 >= 3.5
         * [Anaconda](https://store.continuum.io/cshop/anaconda/)
+        * Must provide the NumPy package; this should come with Anaconda.
```
Comment (moorepants) [inline README.md:509]: Noticed that the Anaconda url above this link gives a 404. This is the correct link: https://www.anaconda.com/distribution/

File: README.md [hunk 2]
```diff
@@ -781,6 +782,7 @@ ctest -j8
         * Mac OSX comes with Python, but you could also use:
         * [`brew install python`](http://brew.sh),
         * [Anaconda](https://store.continuum.io/cshop/anaconda/)
+        * Must provide the NumPy package; this should come with Anaconda.
```
Comment (moorepants) [inline README.md:785]: Same for the line above this one.

File: README.md (hunk 1 above)
Comment (chrisdembia) [reply README.md:509]: Thank you; fixed.

**Comment by aymanhab** [comment]
@chrisdembia this looks exciting, some high level questions:
- numpy site recommends scipy but scipy doesn't appear on ci files, is that intended?
- Does numpy work seamlessly with all variants of python distributions? (we typically support python.org but not Anaconda, Canopy and .the variety of distributions) 
- Would anything here change when we upgrade to swig 4.0 which should happen soon to avoid the issues we keep getting with users building with latest swig? Thank you

**Comment by carmichaelong** [comment]
___
*[Bindings/Python/swig/python_simbody.i, line 193 at r3](https://reviewable.io/reviews/opensim-org/opensim-core/2706#-M4GowdV-goP9XJWKUUa:-M4HyPvnCrSB4PAEM7b_:bbpqx2s) ([raw file](https://github.com/opensim-org/opensim-core/blob/ac24a9499bf45b5b5d7fb65b2ecb657d2e445b24/Bindings/Python/swig/python_simbody.i#L193)):*
<details><summary><i>Previously, chrisdembia (Christopher Dembia) wrote…</i></summary><blockquote>

`to_numpy()` is another option that is somewhat consistent with the interface of the Python `pandas` package.
</blockquote></details>

I think either Mat or something more python specific are both fine as long as documentation/examples exist. Otherwise, i'd vote either for `asarray()` or `to_numpy()`

**Comment by aymanhab** [comment]
to_numpy seems to strike a reasonable balance. 
I assume this has been tested with  some incarnation of Matlab since these are not integrated into ci.
Nice job @chrisdembia 👍

**Comment by chrisdembia** [comment]
> I assume this has been tested with some incarnation of Matlab since these are not integrated into ci.

Yes I tested locally with Matlab.

I will use `to_numpy()`.

**Comment by chrisdembia** [comment]
> * numpy site recommends scipy but scipy doesn't appear on ci files, is that intended?

Yes. NumPy does not depend on SciPy is not necessary. NumPy is the pure matrix package and SciPy is all the fun stuff you can do with the matrices.

> Does numpy work seamlessly with all variants of python distributions? (we typically support python.org but not Anaconda, Canopy and .the variety of distributions)

Yes, NumPy is available with all the popular Python distributions.

> Would anything here change when we upgrade to swig 4.0 which should happen soon to avoid the issues we keep getting with users building with latest swig? Thank you

This is a good question. I have not tested this and I'm not sure. I can look into it but I would prefer that this not hold up the PR.

**Comment by carmichaelong** [comment]
LGTM too. One random failed test, I'm okay with either merging as is or restarting that test just in case.

**Comment by chrisdembia** [comment]
Thanks @aymanhab and @carmichaelong. I restarted the test to give it a shot at passing :)

**Comment by chrisdembia** [comment]
Please feel free to merge if you notice the test succeeds.

**Comment by chrisdembia** [comment]
Wooohooo! :) Thank you for merging.

**Pull request merged**